File: src/main/java/org/tjug/versus/fugue/LensExample.java
package org.tjug.versus.fugue;

import io.atlassian.fugue.optic.Lens;
import org.tjug.versus.data.Approve;
import org.tjug.versus.data.Review;
import org.tjug.versus.data.ReviewStatus;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

public class LensExample {


    public static void main(String[] args) {

        Lens<Approve, Integer> approveValueLens = Lens.lens(Approve::getApproveValue,
                newApproveValue -> oldApprove -> new Approve(oldApprove.getUserId(), newApproveValue, oldApprove.getGrantDate())
        );

        Lens<List<Approve>, Approve> firstApproveLens = Lens.lens(list -> list.get(0),
                newApprove -> oldApproveList -> {
                    List<Approve> result = new ArrayList<>();
                    result.add(newApprove);
                    result.addAll(oldApproveList.subList(1, oldApproveList.size()));
                    return result;
                }
        );

        Lens<Review, List<Approve>> reviewApprovesLens = Lens.lens(Review::getApproveHistory,
                approveHistory -> review ->
                        new Review(review.getReviewStatus(),
                                approveHistory,
                                review.getSubmitDate(),
                                review.getChangeId(),
                                review.getProject(),
                                review.getCloseDate(),
                                review.getFiles())
        );

        Lens<List<Approve>, Integer> firstApproveValueLens = firstApproveLens.composeLens(approveValueLens);

        Lens<Review, Integer> reviewApproveValueLens = reviewApprovesLens.composeLens(firstApproveValueLens);

        Approve approve = new Approve(1, -1, new Date());

        Review review = new Review(ReviewStatus.MERGED,
                Arrays.asList(approve),
                new Date(),
                1,
                "project",
                new Date(),
                new ArrayList<>()
                );

        Review newReview = reviewApproveValueLens.set(2).apply(review);
        System.out.println(review.getApproveHistory().get(0).getApproveValue());
        System.out.println(newReview.getApproveHistory().get(0).getApproveValue());


    }
}

File: src/main/scala/org/tjug/versus/scalaz/LensExample.scala
package org.tjug.versus.scalaz

import java.util.{Arrays, Date}

import scala.collection.JavaConverters._

import org.tjug.versus.data.{ReviewStatus, Review, Approve}

import scalaz.Lens

class LensExample {

  def main(args: Array[String]) {

    val approveValueLens = Lens.lensu[Approve, Integer](
      (approve, newValue) => new Approve(approve.getUserId, newValue, approve.getGrantDate),
      approve => approve.getApproveValue
    )

    val firstApproveLens = Lens.lensu[List[Approve], Approve](
      (oldApproveList, newApprove) => newApprove :: oldApproveList.drop(1)
      ,
      list => list(0)
    )

    val reviewApprovesLens = Lens.lensu[Review, List[Approve]](
      (review, newApproveHistory) => new Review(
        review.getReviewStatus(),
        newApproveHistory.asJava,
        review.getSubmitDate(),
        review.getChangeId(),
        review.getProject(),
        review.getCloseDate(),
        review.getFiles()
      ),
      review => review.getApproveHistory().asScala.toList
    )

    val reviewApproveValueLens = approveValueLens.compose(firstApproveLens.compose(reviewApprovesLens))


    val approve: Approve = new Approve(1, -1, new Date)

    val review: Review = new Review(ReviewStatus.MERGED, Arrays.asList(approve), new Date, 1, "project", new Date, Arrays.asList())

    println(reviewApproveValueLens.get(review))
    val newReview = reviewApproveValueLens.set(review, 2)
    println(reviewApproveValueLens.get(newReview))
  }

}
